Implement update for MyQ API v5

nsnyder opened the pull request:
Hey, I know the README isn't totally updated, but I wanted to see about getting a pull request started, so that other people can use this library for their use cases. I don't have everything working _exactly_ the same, but for the most part, it should be pretty small changes. Let me know what you think, and I'll keep working on getting the README updated.

nsnyder commented:
Hey, I think the readme is appropriately updated now. I've also made some changes for this to work with 5.1.

identd113 commented:
Does the 5.1 API support door locks?

nsnyder commented:
@identd113 I don't have a door lock to test with, but I suspect it would.

For one of my projects, I have a `package.json` that looks like this:
```json
{
  "name": "project-name",
  "version": "0.0.1",
  "dependencies": {
    "some-package" : "1.0.x",
    "myq-api" : "git://github.com/nsnyder/myq-api#branch-for-direct-install"
  }
}
```

I suspect something like this would work for you:
```js
var MyQ = require('myq-api');

var account = new MyQ('email', 'password');

var data = {};
console.log('Logging in.');
account.login()
  .then(function (result) {
    console.log('Login result:', result);
    if (result.returnCode !== 0) {
      throw new Error('Login unsuccessful.');
    }

    // Don't know what the device type would be, so don't specify one.
    return account.getDevices();
  })
  .then(result => {
    console.log('Get Devices result: ', result);
    if (result.returnCode !== 0) {
      throw new Error('Failed to find devices.');
    }

    const desiredDeviceName = 'My Lock Name';
    const device = result.devices.find(device => device.name === desiredDeviceName);
    if (!device) {
      throw new Error(`\nCould not find device with name: ${desiredDeviceName}`);
    }

    // This is the part I'm least sure about. You may have to play around with this.
    // I'd try: "lock", "unlock", "locked", "unlocked" to see if you could get it working.
    console.log("Setting lock state.");
    return account.setDeviceState(device.serialNumber, "lock");
  })
  .then(function (result) {
    if (result.returnCode !== 0) {
      throw new Error('setDeviceState!');
    }
    
    console.log("Lock successfully updated!");
  })
  .then(function (result) {
    console.log('setDoorOpen result:', result);
    if (result.returnCode !== 0) {
      throw new Error('setDoorOpen unsuccessful!');
    }
  })
  .catch(function (err) {
    console.error(err);
  });
```

I'd love to get the `type` from the `getDevices` call. (i.e. `type: 'virtualgaragedooropener',`), as well as which commands ("lock", "unlock", "locked", "unlocked" etc.) end up working so that I can add appropriate commands/constants/etc. to make this work a little bit better.

**Make sure you don't reveal your serial number in anything you post here.**

kn0ll commented:
@nsnyder thanks for your work on this. with a little bit of finagling i was able to test your package against my local smart home setup. i have devices of type: `hub`, `wifigaragedooropener`, and `wifigdogateway`.

nsnyder commented:
@kn0ll Thanks for the report! I've updated this branch to include those types. Specifying types while getting devices isn't strictly necessary, but I'll do my best to keep the list up to date as I get reports back :)

@thomasmunduchira I've gone through the Linting issues, updated some documentation, and updated the `package.json` to match. Haven't thoroughly tested it, but I _think_ everything should be working as expected. Let me know if you see any issues that would prevent merging.

kn0ll commented:
@thomasmunduchira i apologize because i know this is not exactly the place for this question, but it's the closest place i'm able to find to ask. despite the readme suggesting to use the issues tab, i don't see one here anymore. :)

is there any interest in adding TypeScript support? you could either provide this natively (using TypeScript yourself), or through generated `.d.ts` files that get pushed up to NPM. my 2 cents... after seeing where you're headed with `constants.js`, i think using TypeScript natively might be a nice win for you guys,

1. are TypeScript types you are interested in a contribution for, either via source, or via `.d.ts` generation?

2. rather nitpicky, but while thinking about TypeScript definitions, i was just wondering: why put all these constants in a namespace / object? it seems like it could be simpler to use the module itself as the namespace? for instance:

```constants.js
export const authBase = {};
export const errorMessages = {};
export const doorCommands = {};
```

nsnyder commented:
@kn0ll Those all sound like good improvements to me. I'll admit, while I think I've got a decent handle on JavaScript in general, the organization of said JavaScript is probably a weak point for myself.

I don't really see any drawbacks to those suggestions, and given some of the changes that have already been made for v5, I think most people are going to have broken scripts if/when they upgrade. I'm not really sure what the best way to integrate your suggestions into this repo are, but I think you could either submit a PR to my branch, (and it could merge in with that), or you could wait until this gets merged and make the changes directly off of master.

Obviously, the more structural changes could be made relatively easily, so I may take a look at doing those.

In terms of backwards compatibility, I'm not sure I know what needs to be done to accomplish that. I think people are probably just going to have to adapt their software to accommodate these changes. Their stuff has probably been broken for quite some time anyway.

thomasmunduchira commented on src/myq.js:
```diff
@@ -3,287 +3,351 @@ const axios = require('axios');
-        return returnError(15);
+    // TODO: Validate device types when we have a more complete list.
+    Object.values(deviceTypes).forEach(deviceType => {
+      if (!Object.values(constants.allDeviceTypes).includes(deviceType)) {
+        // return ErrorHandler.prototype.returnError(15);
       }
-    }
+    });
```
Comment out the for loop so it doesn't run.

nsnyder replied:
Removed entirely this per your suggestions.

thomasmunduchira commented on src/myq.js:
```diff
@@ -3,287 +3,351 @@ const axios = require('axios');
+    return promise
+      .then(() =>
+        this.executeRequest(
+          `${constants.routes.setDevice
+            .replace('{accountId}', this.accountId)
+            .replace('{serialNumber}', serialNumber)}`,
```
No need to use template string.

thomasmunduchira commented:
@kn0ll, I've addressed your questions below:

> are TypeScript types something you are interested in a contribution for, either via source, or via .d.ts generation? i'd be happy to start contributing with some guidance, if you all are interested. if you're not interested in either, i can always go the DefinitelyTyped route.., but that is likely to get out of sync without community support.

TypeScript support would be a fantastic addition! If you were to make such a contribution, I would wait until this gets merged in and a new update is released. At that point, the source code should be stable enough for adding TypeScript support. The extent to which you want to add support for this (rewriting in TypeScript vs just including `.d.ts` files) is up to you - both sound good to me.

> rather nitpicky, but while thinking about TypeScript definitions, i was just wondering: why put all these constants in a namespace / object? it seems like it could be simpler (and actually more robust) to use the module itself as the namespace? for instance:

This might improve working with these constants internally, but it makes it harder to export these constants out of the library and/or be imported by end-users. I also don't have the utmost confidence in these constants and expect the keys _and_ values to change over time, so surfacing them as top-level entities seems out of scope.

> similar nitpicky, why not continue to provide a default a export as an option in v5 (in addition to, not instead of, your current changes)? import myQ from 'myq-api'; (on that note, is there any chance of providing backwards compatibility? i totally don't need this, but i assume it might make someone out there happy, if that's worth it :) )

Users still need to migrate their client-side code in order to support v5 (for example, passing in the serial number instead of the ID to the API) even if we do our best to minimize end-user impact. So I think we have the flexibility to update a few more of these interfaces and extensively document these changes so that clients can get realigned quickly.

nsnyder commented:
@thomasmunduchira from what I'm seeing, I agree with all of your proposals.

I think one other method we could _potentially_ benefit from would be surfacing the option to address devices by their name, rather than their serial number. While people are going to have to migrate from their old code to V5, it seems likely that we'll always be able to get a name, and if MyQ changes from serial number to some other field, any code that checks the name instead would probably be able to continue working without a change to the interface of this package.

I'll address your comments in a bit. It sounds like you've started working on changes to implement some of your ideas, so I can leave you to that, but if there's any of it you want a hand with, let me know and I'll grab it.

nsnyder commented on src/myq.js:
```diff
@@ -3,287 +3,345 @@ const axios = require('axios');
+    return errorhandler.prototype.returnerror(14, null, data);
+  }
+  if (status === 404) {
+    if (data.code === '404.401') {
```
I _think_ the backend changed its response, but I can't be certain. I'll change this bit and include a comment.

nsnyder commented on src/myq.js:
```diff
@@ -3,287 +3,345 @@ const axios = require('axios');
 
-        const { data } = response;
+          const { data } = response;
+          switch (response.status) {
```
Nope, just accounting for the responses I'd seen. I'll clean this up.

nsnyder commented on src/myq.js:
```diff
@@ -3,287 +3,345 @@ const axios = require('axios');
+  }
+
+  getDevices(deviceTypeParams) {
+    let promise = Promise.resolve(() => null);
```
I think I've addressed this properly, but let me know if I misunderstood this comment.

nsnyder commented on src/myq.js:
```diff
@@ -3,287 +3,345 @@ const axios = require('axios');
 
-        const { data } = response;
+        const { response: { data, status } } = returnValue;
+        if (![200, 204].includes(status)) {
```
Removed this.

nsnyder commented on src/myq.js:
```diff
@@ -3,287 +3,351 @@ const axios = require('axios');
+  getAccountInfo() {
+    return this.executeRequest(constants.routes.account, 'get', { expand: 'account' })
+      .then(returnValue => {
+        if (returnValue.returnCode !== 0 && typeof returnValue.returnCode !== 'undefined') {
```
No good reason. Removing this.

nsnyder commented on src/constants.js:
```diff
@@ -19,10 +35,26 @@ const constants = {
     9: 'not closed',
   },
+  lightCommands: {
+    on: 'on',
+    off: 'off',
```
I'll switch these then. I don't actually have a light to test with.

nsnyder commented on src/myq.js:
```diff
@@ -3,287 +3,351 @@ const axios = require('axios');
+
+  executeRequest(route, method, params, data) {
+    const isLoginRequest = route === constants.routes.login;
+    const headers = {
+      'Content-Type': 'application/json',
+      MyQApplicationId: constants.appId,
+    };
```
Updated. Good to keep an eye on the user agent. To this point, I haven't had any issues related to that with my garage door, but I'll keep an eye out.

nsnyder replied:
I think [this comment](https://github.com/arraylabs/pymyq/issues/46#issuecomment-651913147) is notable. We may decide it's best not to send anything for the time being.

nsnyder commented:
@thomasmunduchira Done! Thanks for the critical eye on this, I appreciate it.

I'll be happy to test it once it's ready. I've only got a garage door, but I've been running it off my own git branch for the time being. It'll be nice to have it pointed at this package instead :)

nsnyder commented:
@thomasmunduchira Conflicts resolved and squashed!

thomasmunduchira commented:
Merged in! As mentioned before, I will ping this thread before the new npm module is rolled out. Thanks again @nsnyder!

JamsheedMistri commented:
@thomasmunduchira Any ETA for the new npm module being rolled out?

bsell93 commented:
There is an issue with the merged code. I have a fix in #10

bsell93 commented:
Also added back in the ability to pass types to `getDevices` in #11

nsnyder commented:
@bsell93 Good catch on the token. PR #10 should be merged. As for `getDevices`, per [this discussion](https://github.com/thomasmunduchira/myq-api/pull/3#pullrequestreview-444441644), I think we're going to avoid that functionality in this implementation.

kn0ll commented:
thank you guys for all your work on this!!

bsell93 commented:
@nsnyder Ah, thanks for pointing me there. I will close #11. 😄

thomasmunduchira commented:
#10 has been merged! Thanks for the contribution @bsell93 :)

@JamsheedMistri I apologize for the delay! I'm taking the time to deep dive into the API and align usage patterns with JavaScript conventions, add tests, etc. I'm aiming to have a preliminary version out this weekend!

thomasmunduchira commented:
I cut a new release to the npm registry: [mpq-api](https://www.npmjs.com/package/myq-api). Happy hacking everyone!

nsnyder commented:
Excellent! Looking forward to giving it a spin. I've been crazy-busy as of late, but I'll be sure to get using it in the not to near future and let you know if I run into any issues :)

nsnyder commented:
Finally got caught up enough to update my existing scripts to use the new API. Working wonderfully!
